README
**********************************************************************
 libaec - Adaptive Entropy Coding library
**********************************************************************

Libaec provides fast lossless compression of 1 up to 32 bit wide
signed or unsigned integers (samples). The library achieves best
results for low entropy data as often encountered in space imaging
instrument data or numerical model output from weather or climate
simulations. While floating point representations are not directly
supported, they can also be efficiently coded by grouping exponents
and mantissa.

Libaec implements Golomb Rice coding as defined in the Space Data
System Standard documents 121.0-B-2 [1] and 120.0-G-2[2].


**********************************************************************
 Patents
**********************************************************************

In doc/license.txt a clarification on potentially applying
intellectual property rights is given.


**********************************************************************
 Installation
**********************************************************************

See INSTALL for details.


**********************************************************************
 Encoding
**********************************************************************

In this context efficiency refers to the size of the encoded
data. Performance refers to the time it takes to encode data.

Suppose you have an array of 32 bit signed integers you want to
compress. The pointer pointing to the data shall be called *source,
output goes into *dest.

#include <libaec.h>

...
    struct aec_stream strm;
    int32_t *source;
    unsigned char *dest;

    /* input data is 32 bits wide */
    strm.bits_per_sample = 32;

    /* define a block size of 16 */
    strm.block_size = 16;

    /* the reference sample interval is set to 128 blocks */
    strm.rsi = 128;

    /* input data is signed and needs to be preprocessed */
    strm.flags = AEC_DATA_SIGNED | AEC_DATA_PREPROCESS;

    /* pointer to input */
    strm.next_in = (unsigned char *)source;

    /* length of input in bytes */
    strm.avail_in = source_length * sizeof(int32_t);

    /* pointer to output buffer */
    strm.next_out = dest;

    /* length of output buffer in bytes */
    strm.avail_out = dest_length;

    /* initialize encoding */
    if (aec_encode_init(&strm) != AEC_OK)
        return 1;

    /* Perform encoding in one call and flush output. */
    /* In this example you must be sure that the output */
    /* buffer is large enough for all compressed output */
    if (aec_encode(&strm, AEC_FLUSH) != AEC_OK)
        return 1;

    /* free all resources used by encoder */
    aec_encode_end(&strm);
...

block_size can vary from 8 to 64 samples. Smaller blocks allow the
compression to adapt more rapidly to changing source
statistics. Larger blocks create less overhead but can be less
efficient if source statistics change across the block.

rsi sets the reference sample interval. A large RSI will improve
performance and efficiency. It will also increase memory requirements
since internal buffering is based on RSI size. A smaller RSI may be
desirable in situations where each RSI will be packetized and possible
error propagation has to be minimized.

Flags:

AEC_DATA_SIGNED: input data are signed integers. Specifying this
correctly increases compression efficiency. Default is unsigned.

AEC_DATA_PREPROCESS: preprocessing input will improve compression
efficiency if data samples are correlated. It will only cost
performance for no gain in efficiency if the data is already
uncorrelated.

AEC_DATA_MSB: input data is stored most significant byte first
i.e. big endian. You have to specify AEC_DATA_MSB even if your host
architecture is big endian. Default is little endian on all
architectures.

AEC_DATA_3BYTE: the 17 to 24 bit input data is stored in three
bytes. This flag has no effect for other sample sizes.

AEC_RESTRICTED: use a restricted set of code options. This option is
only valid for bits_per_sample <= 4.

AEC_PAD_RSI: assume that the encoded RSI is padded to the next byte
boundary while decoding. The preprocessor macro ENABLE_RSI_PADDING
needs to be defined while compiling for the encoder to honour this
flag.

Data size:

Except for the AEC_DATA_3BYTE case for 17 to 24 bit data, the
following rules apply for deducing storage size from sample size
(bits_per_sample):

sample size   storage size
 1 -  8 bits  1 byte
 9 - 16 bits  2 bytes
17 - 32 bits  4 bytes (also for 17 to 24 bits if AEC_DATA_3BYTE is
                       not set)

If a sample requires less bits than the storage size provides, then
you have to make sure that unused bits are not set. Libaec does not
check this for performance reasons and will produce undefined output
if unused bits are set. All input data must be a multiple of the
storage size in bytes. Remaining bytes which do not form a complete
sample will be ignored.

Libaec accesses next_in and next_out buffers only bytewise. There are
no alignment requirements for these buffers.

Flushing:

aec_encode can be used in a streaming fashion by chunking input and
output and specifying AEC_NO_FLUSH. The function will return if either
the input runs empty or the output buffer is full. The calling
function can check avail_in and avail_out to see what occurred. The
last call to aec_encode() must set AEC_FLUSH to drain all
output. aec.c is an example of streaming usage of encoding and
decoding.

Output:

Encoded data will be written to the buffer submitted with
next_out. The length of the compressed data is total_out.

See libaec.h for a detailed description of all relevant structure
members and constants.


**********************************************************************
 Decoding
**********************************************************************

Using decoding is very similar to encoding, only the meaning of input
and output is reversed.

#include <libaec.h>

...
    struct aec_stream strm;
    /* this is now the compressed data */
    unsigned char *source;
    /* here goes the uncompressed result */
    int32_t *dest;

    strm.bits_per_sample = 32;
    strm.block_size = 16;
    strm.rsi = 128;
    strm.flags = AEC_DATA_SIGNED | AEC_DATA_PREPROCESS;
    strm.next_in = source;
    strm.avail_in = source_length;
    strm.next_out = (unsigned char *)dest;
    strm.avail_out = dest_lenth * sizeof(int32_t);
    if (aec_decode_init(&strm) != AEC_OK)
        return 1;
    if (aec_decode(&strm, AEC_FLUSH) != AEC_OK)
        return 1;
    aec_decode_end(&strm);
...

The size of the output buffer (next_out) should be a multiple of the
storage size in bytes. If the buffer is not a multiple of the storage
size and the buffer would be filled to the last sample, the error code
AEC_MEM_ERROR is returned.

It is essential for decoding that parameters like bits_per_sample,
block_size, rsi, and flags are exactly the same as they were for
encoding. Libaec does not store these parameters in the coded stream
so it is up to the calling program to keep the correct parameters
between encoding and decoding.

The actual values of coding parameters are in fact only relevant for
efficiency and performance. Data integrity only depends on consistency
of the parameters.


**********************************************************************
 References
**********************************************************************

[1] Consultative Committee for Space Data Systems. Lossless Data
Compression. Recommendation for Space Data System Standards, CCSDS
121.0-B-2. Blue Book. Issue 2. Washington, D.C.: CCSDS, May 2012.
http://public.ccsds.org/publications/archive/121x0b2.pdf

[2] Consultative Committee for Space Data Systems. Lossless Data
Compression.  Recommendation for Space Data System Standards, CCSDS
120.0-G-3. Green Book. Issue 3. Washington, D.C.: CCSDS, April
2013.
http://public.ccsds.org/publications/archive/120x0g3.pdf


ChangeLog
2016-08-16  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	Better compatibility with OSX for make check

	Allow Cygwin to build DLLs

	v0.3.4

	Pad incomplete last line when in SZ compatibility mode

2016-05-12  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	v0.3.3

	Fixed bug with zero blocks in the last RSI (reference sample
	interval) when data size is not a multiple of RSIs or segments (64
	blocks) and the zero region reaches a sement boundary.

	Vectorization improvement for Intel compiler

	More robust error handling

	Better compatibility with netcdf's build process

2015-02-04  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	v0.3.2

	Allow nonconforming block sizes in SZ mode.

	Performance improvement for decoder.

2014-10-23  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	v0.3.1

	Allow incomplete scanlines in SZ mode.

2014-08-06  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	v0.3

	Performance improvement for encoding pre-precessed data.

	More efficient coding of second extension if reference
	sample is present.

	Port library to Windows (Visual Studio).

	Support building with CMake.

	Benchmarking target using ECHAM data (make bench)


2014-02-12  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	v0.2

	Fixed incorrect length calculation in assessment of Second
	Extension coding.

	Unlimited encoding of fundamental sequences.

	Additional testing with official CCSDS sample data.

	Handle corrupted compressed data more gracefully.

	Support restricted coding options from latest standard.

	Facilitate generation of SIMD instructions by compiler.

2013-05-21  Mathis Rosenhauer  <rosenhauer@dkrz.de>

	Initial relaease: v0.1.


src/aec.c
/**
 * @file aec.c
 *
 * @author Mathis Rosenhauer, Deutsches Klimarechenzentrum
 * @author Moritz Hanke, Deutsches Klimarechenzentrum
 * @author Joerg Behrens, Deutsches Klimarechenzentrum
 * @author Luis Kornblueh, Max-Planck-Institut fuer Meteorologie
 *
 * @section LICENSE
 * Copyright 2012 - 2014
 *
 * Mathis Rosenhauer,                 Luis Kornblueh
 * Moritz Hanke,
 * Joerg Behrens
 *
 * Deutsches Klimarechenzentrum GmbH  Max-Planck-Institut fuer Meteorologie
 * Bundesstr. 45a                     Bundesstr. 53
 * 20146 Hamburg                      20146 Hamburg
 * Germany                            Germany
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above
 *    copyright notice, this list of conditions and the following
 *    disclaimer in the documentation and/or other materials provided
 *    with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @section DESCRIPTION
 *
 * CLI frontend for Adaptive Entropy Coding library
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "libaec.h"

#define CHUNK 10485760

int get_param(unsigned int *param, int *iarg, char *argv[])
{
    if (strlen(argv[*iarg]) == 2) {
        (*iarg)++;
        if (argv[*iarg][0] == '-')
            return 1;
        else
            *param = atoi(argv[*iarg]);
    } else {
        *param = atoi(&argv[*iarg][2]);
    }
    return 0;
}

int main(int argc, char *argv[])
{
    struct aec_stream strm;
    unsigned char *in;
    unsigned char *out;
    size_t total_out;
    unsigned int chunk;
    int status;
    int input_avail, output_avail;
    char *infn, *outfn;
    FILE *infp, *outfp;
    int dflag;
    char *opt;
    int iarg;

    chunk = CHUNK;
    strm.bits_per_sample = 8;
    strm.block_size = 8;
    strm.rsi = 2;
    strm.flags = AEC_DATA_PREPROCESS;
    dflag = 0;
    iarg = 1;

    while (iarg < argc - 2) {
        opt = argv[iarg];
        if (opt[0] != '-')
            goto FAIL;
        switch (opt[1]) {
        case '3':
            strm.flags |= AEC_DATA_3BYTE;
            break;
        case 'N':
            strm.flags &= ~AEC_DATA_PREPROCESS;
            break;
        case 'b':
            if (get_param(&chunk, &iarg, argv))
                goto FAIL;
            break;
        case 'd':
            dflag = 1;
            break;
        case 'j':
            if (get_param(&strm.block_size, &iarg, argv))
                goto FAIL;
            break;
        case 'm':
            strm.flags |= AEC_DATA_MSB;
            break;
        case 'n':
            if (get_param(&strm.bits_per_sample, &iarg, argv))
                goto FAIL;
            break;
        case 'p':
            strm.flags |= AEC_PAD_RSI;
            break;
        case 'r':
            if (get_param(&strm.rsi, &iarg, argv))
                goto FAIL;
            break;
        case 's':
            strm.flags |= AEC_DATA_SIGNED;
            break;
        case 't':
            strm.flags |= AEC_RESTRICTED;
            break;
        default:
            goto FAIL;
        }
        iarg++;
    }

    if (argc - iarg < 2)
        goto FAIL;

    infn = argv[iarg];
    outfn = argv[iarg + 1];

    if (strm.bits_per_sample > 16) {
        if (strm.bits_per_sample <= 24 && strm.flags & AEC_DATA_3BYTE)
            chunk *= 3;
        else
            chunk *= 4;
    } else if (strm.bits_per_sample > 8) {
        chunk *= 2;
    }

    out = (unsigned char *)malloc(chunk);
    in = (unsigned char *)malloc(chunk);

    if (in == NULL || out == NULL)
        exit(-1);

    total_out = 0;
    strm.avail_in = 0;
    strm.avail_out = chunk;
    strm.next_out = out;

    input_avail = 1;
    output_avail = 1;

    if ((infp = fopen(infn, "rb")) == NULL) {
        fprintf(stderr, "ERROR: cannot open input file %s\n", infn);
        return 1;
    }
    if ((outfp = fopen(outfn, "wb")) == NULL) {
        fprintf(stderr, "ERROR: cannot open output file %s\n", infn);
        return 1;
    }

    if (dflag)
        status = aec_decode_init(&strm);
    else
        status = aec_encode_init(&strm);

    if (status != AEC_OK) {
        fprintf(stderr, "ERROR: initialization failed (%d)\n", status);
        return 1;
    }

    while(input_avail || output_avail) {
        if (strm.avail_in == 0 && input_avail) {
            strm.avail_in = fread(in, 1, chunk, infp);
            if (strm.avail_in != chunk)
                input_avail = 0;
            strm.next_in = in;
        }

        if (dflag)
            status = aec_decode(&strm, AEC_NO_FLUSH);
        else
            status = aec_encode(&strm, AEC_NO_FLUSH);

        if (status != AEC_OK) {
            fprintf(stderr, "ERROR: %i\n", status);
            return 1;
        }

        if (strm.total_out - total_out > 0) {
            fwrite(out, strm.total_out - total_out, 1, outfp);
            total_out = strm.total_out;
            output_avail = 1;
            strm.next_out = out;
            strm.avail_out = chunk;
        } else {
            output_avail = 0;
        }

    }

    if (dflag) {
        aec_decode_end(&strm);
    } else {
        if ((status = aec_encode(&strm, AEC_FLUSH)) != AEC_OK) {
            fprintf(stderr, "ERROR: while flushing output (%i)\n", status);
            return 1;
        }

        if (strm.total_out - total_out > 0)
            fwrite(out, strm.total_out - total_out, 1, outfp);

        aec_encode_end(&strm);
    }

    fclose(infp);
    fclose(outfp);
    free(in);
    free(out);
    return 0;

FAIL:
    fprintf(stderr, "NAME\n\taec - encode or decode files ");
    fprintf(stderr, "with Adaptive Entropy Coding\n\n");
    fprintf(stderr, "SYNOPSIS\n\taec [OPTION]... SOURCE DEST\n");
    fprintf(stderr, "\nOPTIONS\n");
    fprintf(stderr, "\t-3\n\t\t24 bit samples are stored in 3 bytes\n");
    fprintf(stderr, "\t-N\n\t\tdisable pre/post processing\n");
    fprintf(stderr, "\t-b size\n\t\tinternal buffer size in bytes\n");
    fprintf(stderr, "\t-d\n\t\tdecode SOURCE. If -d is not used: encode.\n");
    fprintf(stderr, "\t-j samples\n\t\tblock size in samples\n");
    fprintf(stderr, "\t-m\n\t\tsamples are MSB first. Default is LSB\n");
    fprintf(stderr, "\t-n bits\n\t\tbits per sample\n");
    fprintf(stderr, "\t-p\n\t\tpad RSI to byte boundary\n");
    fprintf(stderr, "\t-r blocks\n\t\treference sample interval in blocks\n");
    fprintf(stderr, "\t-s\n\t\tsamples are signed. Default is unsigned\n");
    fprintf(stderr, "\t-t\n\t\tuse restricted set of code options\n\n");
    return 1;
}
